### ISSUE TITLE
Bluetooth controller causes freezing in macOS Sierra

### ISSUE THREAD

**Issue by Acetown** [libretro/RetroArch#3660]
Since updating to macOS Sierra, Retroarch has started to freeze up for like 10 seconds whenever I connect my Dual Shock 3 via Bluetooth (no issues with USB), if it's connected beforehand Retroarch instead freezes on boot up. Same thing also happens when I load a game or toggle between fullscreen and windowed mode, basically it seems that whatever happens in the back end whenever "<controller> configured in port #0." is prompted to appear on screen now takes about twenty times as long.

**Comment by twinaphex** [comment]
I'm unfortunately not going to upgrade to macOS Sierra for this. Sounds like a regression on Apple's end if it worked just fine in prior OSX versions.

**Comment by CoBrA2168** [comment]
Just chiming in here to say that I've noticed the same exact side affect after updating to Sierra.  I did not experience this with El Capitan.  I'm using the latest Sierra build (10.12.6) and RetroArch (1.6.7) as of the writing of this comment.

I'm a developer and would be willing to contribute, but I have no experience with RetroArch and my free time is limited so I'm probably not the person for this job.  Unless someone can point me in the right direction, in that case I might be able to help out.

**Comment by joshuaseltzer** [comment]
@orbea yes at least for me it is.  I'm on 10.14.2 and using the latest Metal build of Retroarch.

**Comment by orbea** [comment]
Thanks for the update!

**Comment by joshuaseltzer** [comment]
Sure.  If you want me to do any tests or try to help at all, my offer still stands.  But as it was over a year ago, my free time is still pretty limited but I can try my best to help out!

**Comment by orbea** [comment]
Sorry, I think you need someone that knows more about bluetooth and osx than me, I'm just trying to clean up the issue queue of the over abundance of already fixed issues...

**Comment by Acetown** [comment]
It has not been fixed yet. I'll just add that this issue seems to affect rumble as well.
Whenever you encounter a situation in a game that initiates a rumble Retroarch will lock up in the same way for like 10 seconds, then it will progress a frame or two and lock up again. This basically keeps going for as the rumbling is programmed to last, which can end up taking minutes. 
I've tested this in Mupen64Plus and ParaLLEl, and it only happens if the controller is connected via Bluetooth.

**Comment by orbea** [comment]
Can you try mupen64plus standalone? I suspect this could be the OS driver for the gamepad and not RetroArch.

**Comment by joshuaseltzer** [comment]
@orbea I could give that a shot later, but as mentioned this happens on both N64 cores.

Also food for thought - I also use the same computer and controller with GeForce Now and I experience no lock ups (rumble works).

**Comment by orbea** [comment]
I recall that the rumble code is very similar between both cores.

What is GeForce now? Is it using the same OS driver for the gamepad?

**Comment by joshuaseltzer** [comment]
Yeah, I just connect the controller using the macOS bluetooth menu, load up GeForce Now, and it just works (it's a game streaming service).

**Comment by joshuaseltzer** [comment]
@twinaphex did you ever upgrade past El Capitan to experience this problem?  I started glancing at the codebase today (the PS3 controller input) and had no idea where to begin to try to debug this issue.

Since we've determined that this is somehow related to rumble, I thought maybe I'd take a stab at trying to fix this (or at least disabling the rumble all together so it doesn't lock up like it currently does).

**Comment by twinaphex** [comment]
I'm on Mojave, but I normally just use gamepads through USB.

**Comment by joshuaseltzer** [comment]
@twinaphex Wow, very interesting - I just tested the PS3 controller via USB and indeed, the freezing doesn't occur.  Plus rumble works again.

So it appears that rumble with Bluetooth on MacOS is simply broken (at least with the Dualshock 3 controller).

Would you have any idea of where to start looking? Or know someone who might?

**Comment by twinaphex** [comment]
@joshuaseltzer It might be best to head over to #programming on our Libretro discord and have a chat with both 'coldbrewed' and 'aussiebloke' about this (and tell them that Autechre sent you). Not sure as to their availability.

As for the code one has to delve into, it's a mixture of some Cocoa input code and also the HID input system as well - 

https://github.com/libretro/RetroArch/blob/master/input/drivers_hid/iohidmanager_hid.c

This is the backend driver for the HID interface system for OSX. 

https://github.com/libretro/RetroArch/tree/master/input/connect

These are the separate low-level joypad drivers for the HID system.

This is the Cocoa input driver in general -

https://github.com/libretro/RetroArch/blob/master/input/drivers/cocoa_input.c

And then there is also an event callback function defined for Cocoa input handling here (see sendEvent etc.) -

https://github.com/libretro/RetroArch/blob/master/ui/drivers/ui_cocoa.m

**Comment by joshuaseltzer** [comment]
Bumping an old issue here, but I'd like to chime in and say that this is still an issue on the latest version of RetroArch (v1.8.7 at the time of this comment) running on MacOS Mojave 10.14.6 (the latest version of 10.14).

As mentioned in the comments years ago above, this all started when upgrading to Sierra.  I didn't notice any sort of freezing/lagging prior to Sierra.  The freezing occurs when connecting the controller (e.g. initial connection via Bluetooth), when the game starts, when the game is closed, or anytime rumble is enabled and is supposed to occur.  With that said, I'm fairly positive this is somehow related to the rumble logic.

I also did try to set up a dev environment to debug this myself over a year ago, but was unable to get a working environment or have enough time to successfully debug the problem.

**Comment by Sanaki** [comment]
You may want to look over #10408. It sounds like the same issue, and if so I believe it's related to retroarch's attempts to read the controller's battery level.

**Comment by joshuaseltzer** [comment]
@Sanaki interesting - yes that could very well be what's happening.

I wonder if there's a way to enable debug logging on the application.  Would be happy to run a test build or something and look through logs to see where it's hanging.  That could maybe help us determine if it's related to reading the controller's battery or something else entirely.

**Comment by lzaiats** [comment]
Hey there! Old bugs still affects some users ;) I still have a Catalina up & running due to the lack of PS3 Dualshock / Sixaxis controllers after 10.15 (i guess)...

After some deep investigation, I found the real problem...
On Sierra+ OSes, something (read Apple) made [IOHIDDeviceSetReport](https://github.com/libretro/RetroArch/blob/master/input/drivers_hid/iohidmanager_hid.c#L1095)  and [IOHIDDeviceGetReport](https://github.com/libretro/RetroArch/blob/master/input/drivers_hid/iohidmanager_hid.c#L1110) take so long to complete, causing the hang / freeze...

I don't know why, RetroArch calls [connect_ps3->ds3_init](https://github.com/libretro/RetroArch/blob/master/input/connect/connect_ps3.c#L261) every time you load a content or return to the UI (close content)...

From my point of view, the controllers must be initialized once per run (only on RetroArch start) and then assume its already initialized and never call the initialization procedure anymore (I really don't know if I am missing something assuming this).

All that said, I did a quick fix for this, caching the result from **IOHIDDeviceSetReport** and **IOHIDDeviceGetReport** as follow:

```
// we cache the iohidmanager_set_report return value
static int32_t iohidmanager_set_report_response = -1;
static int32_t iohidmanager_set_report(void *handle, uint8_t report_type, uint8_t report_id, uint8_t *data_buf, size_t size)
{
   // return the cached value on further calls
   if(iohidmanager_set_report_response != -1)
      return iohidmanager_set_report_response;
      
   struct iohidmanager_hid_adapter *adapter =
      (struct iohidmanager_hid_adapter*)handle;
   if (adapter)
      iohidmanager_set_report_response = IOHIDDeviceSetReport(adapter->handle,
            translate_hid_report_type(report_type), report_id,
            data_buf + 1, size - 1);
      return iohidmanager_set_report_response;
   return -1;
}
```
and
```
// we cache the iohidmanager_get_report return value
static int32_t iohidmanager_get_report_response = -1;
static int32_t iohidmanager_get_report(void *handle, uint8_t report_type, uint8_t report_id,
      uint8_t *data_buf, size_t size)
{
   // return the cached value on further calls
   if(iohidmanager_get_report_response != -1)
      return iohidmanager_get_report_response;

   struct iohidmanager_hid_adapter *adapter =
      (struct iohidmanager_hid_adapter*)handle;

   if (adapter)
   {
      CFIndex length = size;
      iohidmanager_get_report_response = IOHIDDeviceGetReport(adapter->handle, translate_hid_report_type(report_type), report_id, data_buf, &length);
      return iohidmanager_get_report_response;
   }

   return -1;
}
```
Please note this is a quick and dirty fix and it's far from ideal, but it **WORKS!** RetroArch hangs / freeze only on startup now... Tested on latest 1.16.0 version ;)

Also, creating a flag **ds3_initialized** on [connect_ps3->ds3_init](https://github.com/libretro/RetroArch/blob/master/input/connect/connect_ps3.c#L261) didn't work and I am too lazy to go down the rabbit hole to find out why...

I hope this fix help some other Dualshock / Sixaxis owners out there!

Cheers!

**Comment by joshuaseltzer** [comment]
@lzaiats I'll have to go back and check myself, but I believe issue #10408 actually fixed this issue on macOS. But maybe I'm mis-remembering, and the long hangs/freezes still occurred. If that's the case, then your hack/fixes could be very helpful.

Ideally a cleaner solution would be put in place to only init the ds3 controller when absolutely necessary. Either way, thanks for chiming in! If I get a moment later on, I'll revisit this issue on my macOS Mojave machine on the latest Retroarch release.

**Comment by hankyuller** [comment]
Same issue with RA 1.15 on High Sierra using a dualshock3 

@lzaiats how can I use your code ?
